pkg/client2/client.go

```diff
@@ -0,0 +1,395 @@
+		return 0, 0, errors.WithStack(err)
+	}
+
+	return acc.GetAccountNumber(), acc.GetSequence(), nil
+}
+
+// QueryBankBalances queries for bank balances owned by wallet
+func (c coreClient) QueryBankBalances(ctx context.Context, address types.Address) (map[string]types.Coin, error) {
+	requestCtx, cancel := context.WithTimeout(ctx, requestTimeout)
+	defer cancel()
+
+	// FIXME (wojtek): support pagination
+	resp, err := c.bankQueryClient.AllBalances(requestCtx, &banktypes.QueryAllBalancesRequest{Address: string(address)})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	balances := map[string]types.Coin{}
+	for _, b := range resp.Balances {
+		coin, err := types.NewCoin(b.Amount.BigInt(), b.Denom)
+		if err != nil {
+			return nil, err
+		}
+		balances[b.Denom] = coin
+	}
+	return balances, nil
+}
+
+// BroadcastResult contains results of transaction broadcast
+type BroadcastResult struct {
+	TxHash  string
+	GasUsed int64
+}
+
+// EstimateGas runs the transaction cost estimation and returns new suggested gas limit,
+// in contrast with the default Cosmos SDK gas estimation logic, this method returns unadjusted gas used.
+func (c coreClient) EstimateGas(ctx context.Context, encodedTx []byte) (int64, error) {
+	requestCtx, cancel := context.WithTimeout(ctx, requestTimeout)
+	defer cancel()
+
+	simRes, err := c.txServiceClient.Simulate(requestCtx, &txtypes.SimulateRequest{
+		TxBytes: encodedTx,
+	})
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to simulate the transaction execution")
+	}
+
+	// usually gas has to be multiplied by some adjustment coefficient: e.g. *1.5
+	// but in this case we return unadjusted, so every module can decide the adjustment value
+	return int64(simRes.GasInfo.GasUsed), nil
+}
+
+// Broadcast sends transaction to chain, ensuring it passeses CheckTx. Doesn't await until Tx is uncluded in block.
+func (c coreClient) Broadcast(ctx context.Context, encodedTx []byte) (BroadcastResult, error) {
+	requestCtx, cancel := context.WithTimeout(ctx, requestTimeout)
+	defer cancel()
+
+	res, err := c.clientCtx.Client.BroadcastTxSync(requestCtx, encodedTx)
+	if err != nil {
+		if errors.Is(err, requestCtx.Err()) {
+			return BroadcastResult{}, errors.WithStack(err)
+		}
+
+		errRes := client.CheckTendermintError(err, encodedTx)
+		if !isTxInMempool(errRes) {
+			return BroadcastResult{}, errors.WithStack(err)
+		}
+	} else if res.Code != 0 {
+		txHash := fmt.Sprintf("%X", tmtypes.Tx(encodedTx).Hash())
+		return BroadcastResult{}, errors.Wrapf(cosmoserrors.New(res.Codespace, res.Code, res.Log),
+			"transaction '%s' failed", txHash)
+	}
+
+	return BroadcastResult{
+		TxHash: res.Hash.String(),
+	}, nil
+}
+
+// AwaitTx awaits until a signed transaction is included in a block
+func (c coreClient) AwaitTx(ctx context.Context, encodedTx []byte) (BroadcastResult, error) {
+	txHashBytes := tmtypes.Tx(encodedTx).Hash()
+	txHash := fmt.Sprintf("%X", txHashBytes)
```

Comment:
This part looks like a pattern for the hash creation, I would recommend to extract a func for it.